### PR TITLE
Mttl 664 raise created person event

### PR DESCRIPTION
**Pull request by evak6979**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/MTTL-664

### *What is the problem we're trying to solve*

Sending an SNS message to be published

### *What changes have we introduced*

Added the sns AWS .net library, along with localstack.

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: PersonApi/V1/Factories/ResponseFactory.cs [hunk 1]
```diff
@@ -2,12 +2,39 @@
...
+                DateTime = DateTime.UtcNow,
+                EntityId = person.Id,
+                Id = Guid.NewGuid(),
+                EventType = "PersonCreatedEvent",
+                Version = "v1",
+                SourceDomain = "Person",
+                SourceSystem = "PersonAPI",
+                User = new User
+                {
+                    Id = Guid.NewGuid(),
```
Comment (mikemoody-amido) [inline PersonApi/V1/Factories/ResponseFactory.cs:32]: Is this property actually needed if we set an arbitrary value in it?

File: PersonApi/V1/Factories/ResponseFactory.cs [hunk 2]
```diff
@@ -2,12 +2,39 @@
...
+            return new PersonSns
+            {
+                CorrelationId = Guid.NewGuid(),
+                DateTime = DateTime.UtcNow,
+                EntityId = person.Id,
+                Id = Guid.NewGuid(),
+                EventType = "PersonCreatedEvent",
+                Version = "v1",
+                SourceDomain = "Person",
+                SourceSystem = "PersonAPI",
```
Comment (mikemoody-amido) [inline PersonApi/V1/Factories/ResponseFactory.cs:29]: Maybe put these strings in a constants file somewhere?

File: PersonApi/V1/Controllers/PersonApiController.cs [hunk 3]
```diff
@@ -52,7 +59,9 @@ public async Task<IActionResult> GetPersonByIdAsync([FromRoute] PersonQueryObjec
         [HttpPost]
         public async Task<IActionResult> PostNewPerson([FromBody] PersonRequestObject personRequestObject)
```
Comment (mikemoody-amido) [inline PersonApi/V1/Controllers/PersonApiController.cs:60]: This method should have `[LogCall(LogLevel.Information)]` on it.

File: PersonApi/V1/Factories/ResponseFactory.cs [hunk 4]
```diff
@@ -2,12 +2,39 @@
...
 using PersonApi.V1.Domain;
 using System;
+using PersonApi.V1.Infrastructure.JWT;
 
 namespace PersonApi.V1.Factories
 {
-    public interface IResponseFactory
+
+
+    public interface ISnsFactory
```
Comment (mikemoody-amido) [inline PersonApi/V1/Factories/ResponseFactory.cs:11]: These Sns classes should be in their own code files.

File: PersonApi/Startup.cs [hunk 5]
```diff
@@ -132,6 +134,8 @@ public void ConfigureServices(IServiceCollection services)
                 var xmlPath = Path.Combine(AppContext.BaseDirectory, xmlFile);
                 if (File.Exists(xmlPath))
                     c.IncludeXmlComments(xmlPath);
+
+                services.Configure<AwsConfiguration>(options => Configuration.GetSection("AWS").Bind(options));
```
Comment (mikemoody-amido) [inline PersonApi/Startup.cs:138]: Is this bit supposed to be within the `AddSwaggerGen` call?

File: PersonApi.Tests/AwsIntegrationTests.cs [hunk 6]
```diff
@@ -59,7 +62,7 @@ public class TableDef
 
 
     [CollectionDefinition("DynamoDb collection", DisableParallelization = true)]
```
Comment (mikemoody-amido) [inline PersonApi.Tests/AwsIntegrationTests.cs:64]: Maybe change the collection name to "Aws collection" (and anywhere that uses it)?

File: PersonApi/appsettings.LocalDevelopment.json [hunk 7]
```diff
@@ -11,6 +11,6 @@
   },
   "DynamoDb": {
     "LocalMode": true,
-    "LocalServiceUrl": "http://localhost:8000"
+    "LocalServiceUrl": "http://localhost:4566"
```
[the remark this answers is not in the document]
Comment (evak6979) [reply PersonApi/appsettings.LocalDevelopment.json:14]: Either or container must be running, but by now I run my tests on powershell. I'll revert this to 8000, it slipped :)

File: PersonApi/V1/UseCase/PostNewPersonUseCase.cs [hunk 8]
```diff
@@ -11,17 +12,24 @@ public class PostNewPersonUseCase : IPostNewPersonUseCase
...
 
-        public async Task<PersonResponseObject> ExecuteAsync(PersonRequestObject personRequestObject)
+        public async Task<PersonResponseObject> ExecuteAsync(PersonRequestObject personRequestObject, Token token)
         {
-            var person = await _gateway.PostNewPersonAsync(personRequestObject).
-                ConfigureAwait(false);
+            var person = await _gateway.PostNewPersonAsync(personRequestObject).ConfigureAwait(false);
+
+            var personSnsMessage = _snsFactory.Create(person, token);
+            await _snsGateway.Publish(personSnsMessage).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (evak6979) [reply PersonApi/V1/UseCase/PostNewPersonUseCase.cs:32]: Good question - @amidomichael ?

File: PersonApi/V1/Gateways/PersonSnsGateway.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            _amazonSimpleNotificationService = amazonSimpleNotificationService;
+            _settings = settings;
+            _logger = logger;
+        }
+
+        public async Task Publish(PersonSns personSns)
+        {
+            string message = JsonConvert.SerializeObject(personSns);
+            var request = new PublishRequest { Message = message, TopicArn = Environment.GetEnvironmentVariable("PersonTopicArn") };
```
[the remark this answers is not in the document]
Comment (evak6979) [reply PersonApi/V1/Gateways/PersonSnsGateway.cs]: I can inject +

File: PersonApi/V1/Factories/ResponseFactory.cs [hunk 10]
```diff
@@ -2,12 +2,39 @@
...
+    }
+
+    public class PersonSnsFactory : ISnsFactory
     {
-        PersonResponseObject ToResponse(Person domain);
+        public PersonSns Create(Person person, Token token)
+        {
+            return new PersonSns
+            {
+                CorrelationId = Guid.NewGuid(),
```
[the remark this answers is not in the document]
Comment (evak6979) [reply PersonApi/V1/Factories/ResponseFactory.cs:22]: Yes, we are. It was what @amidomichael suggested (Unless I misunderstood)

File: PersonApi/V1/Factories/ResponseFactory.cs (hunk 1 above)
Comment (evak6979) [reply PersonApi/V1/Factories/ResponseFactory.cs:32]: User information is coming from the JWT token. It doesn't include an ID (for now), so I am using an arbitrary value until it includes one in the future (I hope)

File: PersonApi/Startup.cs (hunk 5 above)
Comment (evak6979) [reply PersonApi/Startup.cs:138]: I don't think so. This simply provides a strongly typed class for appsettings that can then be injected.

File: PersonApi/V1/Factories/ResponseFactory.cs (hunk 1 above)
Comment (mikemoody-amido) [reply PersonApi/V1/Factories/ResponseFactory.cs]: I'd have just left it empty, but I guess that's just as arbitrary.

**Review by mikemoody-amido — APPROVED** [review]
Approved, though you might need to catch an exception in the PostNewPersonUseCase depending on @amidomichael 's ruling...

**Pull request merged**